add reconcile-sccs command, create missing SCCs on server start

pweil- opened the pull request:
Add a command to reconcile cluster bootstrap SCCs with existing SCCs.  I think this should be considered for a low risk addition.  Still testing it out but first pass seems to work.

@liggitt @deads2k @smarterclayton

liggitt commented on pkg/cmd/admin/policy/reconcile_sccs.go:
```diff
@@ -0,0 +1,241 @@
+const ReconcileSCCRecommendedName = "reconcile-sccs"
+
+type ReconcileSCCOptions struct {
+	Confirmed bool
```
this needs more switches, since this object combines grants and bindings. I would expect to be able to reconcile the following things independently:

1. SCC definitions
2. SCC priorities
3. SCC subjects, additive only
4. SCC subjects, replace

liggitt replied:
the things a cluster admin is likely to change are the priorities and users/groups... we need to make it possible to reset those without making it the default

liggitt replied:
see also `diagnostics/cluster/{roles,rolebindings}.go`

pweil- replied:
> SCC subjects, replace

this one is a little tricky.  We use the ensure method to craft additional SA user names to add to the SCC and replacing the groups/users would squash that.  I'll see if there is a good way to do it.  The other pieces seem easy to add flags for.

liggitt replied:
I mean replacing what they have, in the case where they want to reset to defaults

pweil- replied:
Yep, understood.  The problem is here https://github.com/pweil-/origin/blob/6f86f784806558c188170ac235a994d3482c02f5/pkg/cmd/server/origin/ensure.go#L165-165.  I can't just use what is in the default SCCs from bootstrap because we manipulate it with settings from the MasterConfig when installing defaults.  I'll look for a better way to do it

liggitt replied:
we need to be able to use this command on existing clusters to get SCCs in the shape we expect them. whatever inputs the `ensure` function uses to calculate default SCCs also needs to be inputs to this tool

liggitt replied:
the PV controller service account will be another likely addition to that list

liggitt commented on pkg/cmd/admin/policy/reconcile_sccs.go:
```diff
@@ -0,0 +1,241 @@
+
+// hasGrants ensures all expected grants are in the actual grants.
+func hasGrants(expectedGrants []string, actualGrants []string) bool {
+	grants := sets.NewString(actualGrants...)
```
there are existing `sets.String` functions for checking equality, contains all, etc

pweil- commented:
updated to use set methods and added flags for union and definitions.  Still todo is priorities and getting the full config with additional service accounts

pweil- commented on pkg/cmd/admin/policy/reconcile_sccs.go:
```diff
@@ -0,0 +1,245 @@
+
+	cmd.Flags().BoolVar(&o.Confirmed, "confirm", o.Confirmed, "Specify that cluster SCCs should be modified. Defaults to false, displaying what would be replaced but not actually replacing anything.")
+	cmd.Flags().BoolVar(&o.Definitions, "definitions", o.Definitions, "Reconcile SCC definitions (exluding priorities).")
+	cmd.Flags().BoolVar(&o.Union, "additive-only", o.Union, "Preserves extra users and gruops in the SCC.")
```
typo

pweil- commented:
@liggitt I picked in the minimal commit needed to get priorities in here, you can ignore the upstream commit.  I was able to get away with some small refactorings rather than requiring the master config as input.  

I want to do some more thorough testing tomorrow but I have pretty good unit tests for the compute method.  PTAL

pweil- commented:
rebased since priorities just merged

pweil- commented:
@liggitt updated and squashed

1.  always reconcile definitions
1.  better ns name
1.  pass ns to GetBoostrapSCCAccess and have it add the builder /cc @markturansky 
1.  move back to policy subcommand thanks to :arrow_up: 
1.  have additive-only control unioning groups, users, and preserve existing priorities

Will create a bugzilla to get QE eyes on this

pweil- commented:
[test]

pweil- commented:
I will need to rebase when https://github.com/openshift/origin/pull/5465 goes in, it was just tagged

liggitt commented on pkg/cmd/admin/policy/reconcile_sccs.go:
```diff
@@ -0,0 +1,270 @@
+	sort.Sort(stringSorter(actual.Groups))
+	sort.Sort(stringSorter(actual.Users))
+	sort.Sort(stringSorter(expected.Groups))
+	sort.Sort(stringSorter(expected.Users))
```
`sort.StringSlice(...).Sort()`

pweil- replied:
fixed

liggitt commented:
exercise in `test/cmd/admin.sh`

delete a core scc and make sure reconcile recreates it (when --confirm) is passed
```
oc delete scc/restricted
[ ! "$(oc get scc/restricted)" ]
oadm policy reconcile-sccs
[ ! "$(oc get scc/restricted)" ]
oadm policy reconcile-sccs --confirm
oc get scc/restricted
```

add a custom user to a core scc and make sure reconcile does not remove it:
```
oadm policy add-scc-to-user restricted my-restricted-user
[ "$(oc get scc/restricted -o yaml | grep my-restricted-user)" ]
oadm policy reconcile-sccs --confirm
[ "$(oc get scc/restricted -o yaml | grep my-restricted-user)" ]
```

delete a core user or group from a core scc and make sure reconcile readds it:
```
oadm policy remove-scc-from-group restricted system:authenticated
[ ! "$(oc get scc/restricted -o yaml | grep system:authenticated)" ]
oadm policy reconcile-sccs --confirm
[ "$(oc get scc/restricted -o yaml | grep system:authenticated)" ]
```

pweil- commented:
rebased

@markturansky - please double check the pv controller scc (moved into securitycontextconstraints.go)

@liggitt added your snippets to `test-cmd.sh`, thanks for providing them

liggitt commented on pkg/cmd/admin/policy/reconcile_sccs.go:
```diff
@@ -0,0 +1,258 @@
+  $ %[1]s
+
+  # Display the cluster SCCs that would be modified
+  $ %[1]s
```
duplicate help

pweil- commented:
[test]

liggitt commented:
LGTM

smarterclayton commented:
[merge]
